Trace connection errors in Net::HTTP.start

DocX opened the pull request:
Adding trace around connection errors https://github.com/DataDog/dd-trace-rb/issues/1404

WIP: Having issues with the spec. It works when run isolated. But fails when run whole `miniapp_spec.rb`. I suspect there is some interference with WebMock, but was not able to solve it. Looks like `WebMock.disable!` is reloading the unpatched version, removing the tracer patch. Maybe if we can somehow force the ddtrace patch, that could work?

DocX commented on spec/ddtrace/contrib/http/miniapp_spec.rb:
```diff
@@ -5,24 +5,26 @@
+    after do
+      WebMock.reset!
+      WebMock.disable!
+      Datadog::Contrib::HTTP::Patcher.remove_instance_variable(:@done_once) if Datadog::Contrib::HTTP::Patcher.patched?
```
`WebMock.disable!` restores the original version of `Net::HTTP` that does not contain the DD patch. Originally we started `WebMock.enable!` after the patch was loaded, so when restoring, it would just restore to the DD patched Net::HTTP.

Because the connection error tests need to remove the Webmock completely, it is reloading the "pure" `Net::HTTP`. Somehow it interferes with all this patching.

There may be nicer/cleaner way to do all this, this is what I ended up with that works :)

DocX replied:
It would require to properly understand the order of patching during running those specs. If we can guarantee that we always first patch Datadog before any `WebMock.enable!` for the whole `rspec` runtime, then probably we would not need this to force repatching.

DocX commented on lib/ddtrace/contrib/http/ext.rb:
```diff
@@ -11,6 +11,7 @@ module Ext
         ENV_ANALYTICS_SAMPLE_RATE_OLD = 'DD_HTTP_ANALYTICS_SAMPLE_RATE'.freeze
         SERVICE_NAME = 'net/http'.freeze
         SPAN_REQUEST = 'http.request'.freeze
+        CONNECTION_ERROR_RESOURCE = 'UNKNOWN'.freeze
```
The connection error happens before the user call even calls the `request` (in case of not using the shortcut methods provided by `Net::HTTP`), we don't know what kind of method it is.

Feel free to suggest better resource name for the connection errors spans - this will only be present on error spans

delner commented:
Hey @DocX thanks for sharing this!

I've read through the code, but need some clarification. Why do you want `start` to catch errors and put them in a new span (just for that error) as opposed to adding them to the existing span for that request? Is this because the request hasn't been created yet?

If this is the case, then I think the more appropriate change would be to add a span that wraps `start`, whether that is in addition to the existing "request" span or a replacement to it. It could also have the benefit for users of separating the time to open an HTTP connection vs the time to actually send the request, depending on how it works. Then on this `start` span, you could capture and annotate any errors.

Let me know your thoughts!

DocX commented:
Hi @delner, thanks for looking at this 

> Why do you want start to catch errors and put them in a new span (just for that error) as opposed to adding them to the existing span for that request? Is this because the request hasn't been created yet?

Yes, in some use cases of the `Net::HTTP` api the request may not be created yet. As it first starts the connection, and then you can perform multiple requests inside of the single connection. And the existing trace implementation wraps the span around the requests.

The idea why to create only the error span was to not create additional span if the operation is successful, to not introduce dramatic change in the spans/traces as I was thinking someone may have some metrics and/or alarms configured around it.

But let me know if that is not a concern, or how can we add the span so it does not affect people's existing setup.

> It could also have the benefit for users of separating the time to open an HTTP connection vs the time to actually send the request, depending on how it works

This is potentially good idea. Yep we could just always wrap the `start` with a span with different operation name perhaps? `http.connection_start`? Then it would just work in both cases, error / not error

delner commented:
> I'm not sure if it is good design if this would change based on the usage? I feel like the connection should be always sibling - because that is how it works internally.

Yeah, as a matter of principle, we should never hack a trace for the sake of making it look a certain way. Each span should mirror actual execution/sequence... if `start` is a sibling, that's okay as long as it mirrors the internal behavior.

For the sake of clarity, could you illustrate the possible uses of HTTP and the respective traces (that you're concerned about)? Want to make sure I understand the potential issue here.

DocX commented:
Hello, @delner. Sorry for the delay, but finally I got some time to continue on this :) 

Hi, I've updated the implementation to record the spans correctly for each connect/request.

The original problem I was referring to was that the `request` has recursion in the native implementation (that was hidden by WebMock) - when the connection is not started, it calls `start` and then it calls itself again (`request`).

We had some "deduplication" code to not create double spans there. I think the better way was to just pass to `super` when the connection is not established, and let the Net::HTTP start the connection and call `request` again. This time it will always create single span.

I think it make sense what you wrote to just mirror the internal implementation, which is what it does now I think.

The main problem now is to fix all the specs :) I've updated the `miniapp` spec and added "test http server" to avoid using WebMock (because it hides this important internal behaviour).

Let me know if this approach make sense, and I will update rest of the specs.

DocX commented on lib/ddtrace/contrib/http/instrumentation.rb:
```diff
@@ -29,11 +29,13 @@ def self.after_request(&block)
 
           # :yield: +response+
           def request(req, body = nil, &block)
+            return super(req, body, &block) unless started?
```
The internal implementation makes starts the connection and call `request` in recursion if the connection is not yet started.

This would mean we would create two request spans for every request call. This guard make sure it does not happen - it pass to super when connection is not started. This will do what is described above, and will pass to `request` method again, this time with started connection.

DocX commented on spec/ddtrace/contrib/http/request_spec.rb:
```diff
@@ -424,22 +430,23 @@ def expect_request_without_distributed_headers
       let(:span) { spans.first }
+      let(:expected_error) { "Failed to open TCP connection to #{host}:#{port} (#{custom_error_message})" }
 
-      before { stub_request(:get, "#{uri}#{path}").to_raise(custom_error) }
```
I am not sure what was the real world use case for testing custom errors from `request`. According to the [documentation](https://ruby-doc.org/stdlib-2.7.0/libdoc/net/http/rdoc/Net/HTTP.html#method-i-request) it states this on the request methods:

> This method never raises Net::* exceptions.

I also tried to simulate some malformed HTTP response from the server but I did not manage to get exception.

So instead I've changed this to exception on socket open.

This is actually connected to the issue we are trying to solve here. Because the original `request` span was also encapsulating the connection opening, in real world those request spans could have exceptions. But since we split the connection spans out of the request spans, the request span will likely never have exceptions (at least if I get the documentation correctly)
